Catalogue epic 2 3

Kolesnikov-Eugene opened the pull request:
Ссылка на командный таск-трекер:
https://github.com/users/Kolesnikov-Eugene/projects/2/views/1?layout=board

Архитектура проекта: MVVM
Верстка UI: в коде
Декомпозиция проекта в CatalogDecomposition.md в корне проекта

В работе API наблюдаются периодические сбои. Периодически сервер присылает ошибку. Чаще всего ошибки наблюдаются на экране коллекции каталога.

Во второй трети реализован UI второго экрана каталога коллекции nft, а также наполнение данными коллекции с сервера.
 
Обработка лайка, добавление в корзину и переход на страницу автора будет реализовано в последнем эпике.

Kolesnikov-Eugene committed:
CatalogCollectionView UI

Kolesnikov-Eugene committed:
user interaction disabled/enabled

Kolesnikov-Eugene committed:
CatalogCollection screen and CatalogCollectionCell UI ready

Kolesnikov-Eugene committed:
swiftLint desabled
user sandboxing - no

Kolesnikov-Eugene committed:
code clean up

Kolesnikov-Eugene committed:
CatalogCollection image download.
NftService to CatalogCollectionViewModel

Kolesnikov-Eugene committed:
Merge branch 'catalogue_epic_2' of https://github.com/Kolesnikov-Eugene/iOS-FakeNFT-StarterProject-Public into catalogue_epic_2_3

# Conflicts:
#	FakeNFT.xcodeproj/project.pbxproj
#	FakeNFT/Services/Requests/RequestConstants.swift

Kolesnikov-Eugene committed:
swiftLint enabled
UIView+Flash extension
nftService implementation

Kolesnikov-Eugene committed:
loadNftForCollection -> CatalogCollectionView UI update.
fetchAuthorProfile method -> author name update.
refactoring

Kolesnikov-Eugene committed:
authorLinkAnimationView

Kolesnikov-Eugene committed:
likeButton in nftCell.
Error Alert if author page link or nfts url array has not been loaded.

Kolesnikov-Eugene committed:
TestCatalogVC deleted
refactoring

Kolesnikov-Eugene committed:
swiftgen intagrated.
Constants removed.

Kolesnikov-Eugene committed:
network error alert multiple presentation bug fixed

Azonaz commented on FakeNFT/Scenes /Catalog/CatalogViewCell/CatalogTableViewCell.swift:
```diff
@@ -9,102 +9,106 @@ import UIKit
+    // MARK: - Private properties
     private let NFTImageView: UIImageView = {
         let imageView = UIImageView()
-        
```
Я бы изменила на nftImageView, чтобы имя переменной начиналась с маленькой буквы, а не с заглавной.

Kolesnikov-Eugene replied:
Точно! Упустил из виду

Azonaz commented on FakeNFT/Services/CatalogServices/CatalogService.swift:
```diff
@@ -0,0 +1,59 @@
+                            completion(.success((catalog)))
+                        case .failure(let error):
+                            completion(.failure((error)))
+                        }
```
Здесь, кажется, можно убрать лишние скобки, на функциональность не влияет: 
        completion(.success(catalog))
case .failure(let error):
        completion(.failure(error))

Kolesnikov-Eugene replied:
Ага, многовато скобок получилось) 
исправил

Azonaz commented on FakeNFT/Scenes /Catalog/CatalogCollectionView/CatalogCollectionView.swift:
```diff
@@ -0,0 +1,340 @@
+        contentView.addSubview(authorPageLinkButton)
+        contentView.addSubview(catalogDescriptionLabel)
+        contentView.addSubview(collectionView)
+
```
Можно немножко сократить код с .forEach: 
[collectionCoverImageView, catalogNameLabel ...].forEach {
   $0.translatesAutoresizingMaskIntoConstraints = false
   addSubview($0)
}

Kolesnikov-Eugene committed:
Dinamic nftCell width calculation.
CatalogDecomposition update.
Refactoring

Kolesnikov-Eugene committed:
swiftlint.yml
refactoring

Kolesnikov-Eugene committed:
nftImageView rename

Kolesnikov-Eugene committed:
swiftGen autocorrect

Kolesnikov-Eugene committed:
weak reference to cell from CatalogCollectionView.
animation cell view alpha changed to 0.4.
scrollView frameLayoutGuide constraints.

Helvetii951 commented on FakeNFT/DesignSystem/UIView+Flash.swift:
```diff
@@ -0,0 +1,21 @@
+        flash.autoreverses = true
+        flash.repeatCount = .infinity
+        layer.add(flash, forKey: nil)
+    }
```
Желательно, чтобы название функций отражали действие. 

```
func addfFashLayer()
```

Helvetii951 commented on FakeNFT/Generated/Strings.swift:
```diff
@@ -0,0 +1,92 @@
+// swiftlint:disable all
+// Generated using SwiftGen — https://github.com/SwiftGen/SwiftGen
+
+import Foundation
+
+// swiftlint:disable superfluous_disable_command file_length implicit_return prefer_self_in_static_references
+
+// MARK: - Strings
+
+// swiftlint:disable explicit_type_interface function_parameter_count identifier_name line_length
+// swiftlint:disable nesting type_body_length type_name vertical_whitespace_opening_braces
+internal enum L10n {
+  internal enum Catalog {
+    /// Открыть Nft
+    internal static let openNft = L10n.tr("Localizable", "Catalog.openNft", fallback: "Открыть Nft")
+  }
+  internal enum CatalogCell {
+    /// CatalogTableViewCell
+    internal static let reuseIdentifier = L10n.tr("Localizable", "CatalogCell.reuseIdentifier", fallback: "CatalogTableViewCell")
+  }
+  internal enum CatalogCollection {
+    /// Автор коллекции:
+    internal static let authorLabel = L10n.tr("Localizable", "CatalogCollection.authorLabel", fallback: "Автор коллекции:")
+  }
+  internal enum CatalogCollectionCell {
+    /// CatalogCollectionCell
+    internal static let reuseIdentifier = L10n.tr("Localizable", "CatalogCollectionCell.reuseIdentifier", fallback: "CatalogCollectionCell")
+  }
+  internal enum CatalogFilterStorage {
+    /// catalogFilter
+    internal static let key = L10n.tr("Localizable", "CatalogFilterStorage.key", fallback: "catalogFilter")
+  }
```
Опять же. Лучше не хранить сгенеренные файлы в репозитории. Их лучше выносить в .gitignore.

Helvetii951 commented on FakeNFT/Scenes /Catalog/CatalogCollectionView/CatalogCollectionView.swift:
```diff
@@ -0,0 +1,348 @@
+    private let viewModel: CatalogCollectionViewModelProtocol
+    private lazy var backButton: UIButton = {
+        let button = UIButton()
+
```
Совсем не обязательно локализовать `reuseIdentifier`. Лучше перенести в саму ячейку и сделать просто английской строкой.

Helvetii951 commented on FakeNFT/Scenes /Catalog/CatalogCollectionView/CatalogCollectionView.swift:
```diff
@@ -0,0 +1,348 @@
+        view.translatesAutoresizingMaskIntoConstraints = false
+
+        return view
+    }()
```
Рекомендую использовать [ленивые свойства ](https://abhimuralidharan.medium.com/lazy-var-in-ios-swift-96c75cb8a13a)для объявления подобных.  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память.

Helvetii951 commented on FakeNFT/Scenes /Catalog/CatalogCollectionView/CatalogCollectionView.swift:
```diff
@@ -0,0 +1,348 @@
+
+    // MARK: - private methods
+    private func bind() {
+        viewModel.nftsLoaderPuublisher.receive(on: DispatchQueue.main)
+            .sink { [weak self] _ in
+                guard let self = self else { return }
+
+                collectionView.reloadData()
+
+            }.store(in: &subscribes)
```
В таких случаях не обязательно распаковывать `self`, а использовать [optional chaining](https://www.geeksforgeeks.org/optional-chaining-in-swift/).

```
self?.collectionView.reloadData()
```

Helvetii951 commented on FakeNFT/Scenes /Catalog/CatalogCollectionView/CatalogCollectionView.swift:
```diff
@@ -0,0 +1,348 @@
+    private func startAnimation() {
+        authorLinkAnimationView.isHidden = false
+        authorLinkAnimationView.flash()
+    }
```
Если  вызывать`startAnimation()`, в `authorLinkAnimationView.flash()` каждый раз будет добавлен анимированый слой. Было бы неплохо предусмотреть удаление/переиспользование этого слоя.

Helvetii951 commented on FakeNFT/Scenes /Catalog/CatalogCollectionView/CatalogCollectionViewCell.swift:
```diff
@@ -0,0 +1,252 @@
+    private func didSelectRate(gesture: UITapGestureRecognizer) {
+        let location = gesture.location(in: starsContainer)
+        let starWidth = starsContainer.bounds.width / CGFloat(5)
+        let rate = Int(location.x / starWidth) + 1
```
`CGFloat(5)`, `+ 1`. Это - [магические числа.](https://bureau.ru/soviet/20230406/). Желательно выносить эти значения в свойства или структуры с понятным названием.

Helvetii951 commented on FakeNFT/Scenes /Catalog/CatalogView/CatalogView.swift:
```diff
@@ -8,91 +8,93 @@
-    private let reuseIdentifier = Constants.catalogTableViewCellIdentifier
+
+    // MARK: - Private properties
+    private let reuseIdentifier = L10n.CatalogCell.reuseIdentifier
     private let tableView: UITableView = {
         let tableView = UITableView()
-        
```
Опять же. Индикаторы переиспользования не нужно локализировать.

Helvetii951 commented on FakeNFT/Scenes /Catalog/CatalogView/CatalogViewController.swift:
```diff
@@ -72,24 +72,25 @@ final class CatalogViewController: UIViewController {
     func selectedCategory(_ model: Catalog) {
-        let vc = CatalogCollectionViewController()
-        navigationController?.pushViewController(vc, animated: true)
+        let viewController = CatalogCollectionViewController(catalog: model)
+        viewController.modalPresentationStyle = .fullScreen
+        present(viewController, animated: true)
```
Лучше пушить контроллер, а не презентить.

Kolesnikov-Eugene committed:
refactoring after review
